src/puppetlabs/puppetdb/query/events.clj

```diff
@@ -50,49 +50,74 @@
 (defn with-latest-events
   "CTE to wrap unioned queries when distinct_resources is used"
   [query]
-  (format
-    "WITH latest_events AS
-           (SELECT certnames.certname,
-                   resource_events.resource_type,
-                   resource_events.resource_title,
-                   resource_events.property,
-                   MAX(resource_events.timestamp) AS timestamp
-                   FROM resource_events
-                   JOIN certnames ON resource_events.certname_id = certnames.id
-                   WHERE resource_events.timestamp >= ?
-                   AND resource_events.timestamp <= ?
-                   GROUP BY certname, resource_type, resource_title, property)
-           %s" query))
+  (let [collate (if (sutils/postgres?) "COLLATE \"C\"" "")]
+    (format
+      "WITH latest_events AS
+       (SELECT certname,
```

Comment:
It appears to be more efficient to use `DISTINCT ON` construct here. See [full](https://gist.github.com/kbrezina/49ea3f93f6348bfb6443#file-resource_events_certname_resource_timestamp-txt) or [reduced](https://gist.github.com/kbrezina/351ae99e67d65dac3dc9) query. Make sure to create [resource_events_certname_resource_timestamp](https://gist.github.com/kbrezina/49ea3f93f6348bfb6443) index first.